### ISSUE TITLE
Spring Magic: The Gathering 

### ISSUE THREAD

**Issue by white-ws** [white-ws/things-i-dont-know#15]
### Constraint annotation

Example of a custom constraint annotation. All the fields below are mandatory to have.

```kotlin
@Constraint(validatedBy = [CustomConstraintValidator::class])
@Target(AnnotationTarget.FIELD, AnnotationTarget.FUNCTION, AnnotationTarget.ANNOTATION_CLASS, AnnotationTarget.VALUE_PARAMETER)
@Retention
annotation class CustomConstraint(
    val message: String = "Custom error message.",
    val groups: Array<KClass<out Any>> = [],
    val payload: Array<KClass<out Payload>> = []
)

```

#### Breakdown

- `@Constraint` marks this annotation to be a Bean Validation constraint. `validatedBy = ` specifies the actual validator implementation of the constraint.

- `@Target` defines the supported target element types for the constraint.
   We have:     
      - `FIELD` : field
      - `FUNCTION` : function
      - `ANNOTATION_CLASS` : annotation class 
      - `VALUE_PARAMETER` : parameter of a function
      ..... and more

- `@Retention` determines how the annotation is stored in binary output, default setting is `RUNTIME` - Annotation is stored in binary output and visible for reflection.

- `val message` is the message that will be displayed when data is invalid

- `val groups` is an array that stores validation group names that this constraint belongs to, [see Grouping Constraints](https://docs.jboss.org/hibernate/validator/5.0/reference/en-US/html/chapter-groups.html#example-drive-away).

- `val payload` is an array that stores metadata information

#### Read more

https://docs.jboss.org/hibernate/validator/5.0/reference/en-US/html/validator-customconstraints.html#validator-customconstraints

#### `@Valid` vs `@Validated` 

After added this annotation to a controller, any request coming to this specific endpoint will have its parameters validated.
tl;dr: not much difference 😩 

- `@Valid` is from JSR-303 standard

- `@Validated` is from Spring standard

### `@ClientHttpRequestInterceptor`
